module_utils/certificate/providers/base.py

```diff
@@ -0,0 +1,291 @@
+
+        The argument `cert_pem` it's a certificate in PEM format. The argument
+        `csr_pem` it's a CSR in PEM format.
+
+        If a CSR is provided it has precedence over the certificate itself. The
+        reason behind that is that a certificate might not have all attributes
+        from the CSR since the CA can choose which attributes to honor. Using the
+        certificate as the primary source of information could lead to idempotency
+        issues.
+
+        If neither CSR or Certificate are available return None instead.
+
+        """
+        # pylint: disable=protected-access
+        if csr_pem:
+            x509_obj = x509.load_pem_x509_csr(csr_pem, default_backend())
+        elif cert_pem:
+            x509_obj = x509.load_pem_x509_certificate(cert_pem, default_backend())
+        else:
+            return None
+
+        cert_like = cls()
+        cert_like.cert_data = cert_like._get_info_from_x509(x509_obj)
+        return cert_like
+
+    def _get_info_from_x509(self, x509_obj):
+        info = {}
+        if not x509_obj:
+            return info
+
+        self._x509_obj = x509_obj
+
+        info["dns"] = self._get_san_values(x509.DNSName)
+        info["ip"] = self._get_san_values(x509.IPAddress)
+        info["common_name"] = self._get_subject_values(NameOID.COMMON_NAME)
+
+        return info
+
+    @property
+    def dns(self):
+        """Return the DNS(s) in the certificate."""
+        return self.cert_data.get("dns")
+
+    @property
+    def ip(self):
+        """Return the IP(s) in the certificate."""
+        return self.cert_data.get("ip")
+
+    @property
+    def common_name(self):
+        """Return the certificate common_name."""
+        return self.cert_data.get("common_name")
+
+    def _get_subject_values(self, oid):
+        values = self._x509_obj.subject.get_attributes_for_oid(oid)
+        if values:
+            return values[0].value
+        return None
+
+    def _get_san_values(self, san_type):
+        if not self._subject_alternative_names:
+            return []
+        return self._subject_alternative_names.value.get_values_for_type(san_type)
+
+    @property
+    def _subject_alternative_names(self):
+        try:
+            san = self._x509_obj.extensions.get_extension_for_oid(
+                x509.oid.ExtensionOID.SUBJECT_ALTERNATIVE_NAME,
+            )
+        except x509.ExtensionNotFound:
+            return []
+        return san
+
+    def __eq__(self, other):
+        """Compare two instances of CertificateProxy for idempotency purposes."""
+        # TODO: compare CA here
+        #   Currently each provider is responsible for implementing the CA
+        #   comparison.
+
+        if not isinstance(other, CertificateProxy):
+            raise TypeError(
+                "Cannot compare 'CertificateProxy' with '{}'".format(type(other),)
+            )
+
+        # Remove empty sequencies and strings, false and None values
+        #   before comparison happens
```

Comment:
"sequences"

I can see removing empty values like `{}` and `[]` - is it possible that `''` empty string is a significant value?  Is it possible that `false` is a significant value?  What about numeric values? e.g. a field with a value of `0` will not be included.

Reply:
That a good question. It's likely that will have cases where None, empty string and False can be meaningful. With the current parameters (just testing parameters) this was not the case.

I was planning to deal with the other cases when I actually find the situation. Is that reasonable?

Reply:
> I was planning to deal with the other cases when I actually find the situation. Is that reasonable?

Yes.

Reply:
Typo fixed in 4e74b07b67c935b8d577375e7779606444905190